### ISSUE TITLE
Helm upgrade does not apply to initContainers

### ISSUE THREAD

**Issue by jaytaylor** [helm/helm#4374]
**TL;DR:**

### initContainer.image values are not updated after `helm upgrade` is executed.

---

See the `kubectl get pod -o yaml` outputs below.  Non-relevant portions have been pruned for brevity.

The interesting parts are the before and after `image: ` values.

**Pre-upgrade Elasticsearch client pod:**

```yaml
apiVersion: v1
items:
- apiVersion: v1
  kind: Pod
  metadata:
    name: elasticsearch-client-2945103334-l5xkk
    namespace: default
  spec:
    containers:
      image: docker-registry.lan:5000/elasticsearch-kubernetes:2.0.0
      imagePullPolicy: Always
      name: elasticsearch-client
    initContainers:
    - command:
      - sysctl
      - -w
      - vm.max_map_count=262144
      image: docker-registry.lan:5000/elasticsearch-kubernetes:2.0.0
      imagePullPolicy: Always
      name: sysctl-vm-max-map-count
```

**Post-upgrade Elasticsearch client pod:**

```yaml
apiVersion: v1
items:
- apiVersion: v1
  kind: Pod
  metadata:
    name: elasticsearch-client-3550306936-2g6m0
    namespace: default
  spec:
    containers:
      image: docker-registry.lan:5000/elasticsearch-kubernetes:2.0.1
      imagePullPolicy: Always
      name: elasticsearch-client
    initContainers:
    - command:
      - sysctl
      - -w
      - vm.max_map_count=262144
      image: docker-registry.lan:5000/elasticsearch-kubernetes:2.0.0
      imagePullPolicy: Always
      name: sysctl-vm-max-map-count
```

Note that after `helm upgrade`, rather than switching to "2.0.1", the initContainer image version has remained stuck at `2.0.0`, even after the pod container image has been updated.

This behavior seems buggy.  After running the idempotent upgrade command (`helm install --upgrade ...`), I expected that the initContainer definitions would be updated.

Additionally, inspecting the output of `--debug --dry-run` shows that the initContainer version will be set to `2.0.1`.  After proceeding to run the upgrade command, it becomes apparent the plan presented by helm is not what ended up being executed.

This is problematic because sometimes mismatched container/initContainer pairs do not mesh together, and the end result is a service interruption or outage.

---

Output of `helm version`:

(N.B. reproduced this issue with the following versions of helm+tiller: `2.8.2`, `2.9.0`, and `2.10.0-rc1`)

```
Client: &version.Version{SemVer:"v2.9.0", GitCommit:"f6025bb9ee7daf9fee0026541c90a6f557a3e0bc", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.9.0", GitCommit:"f6025bb9ee7daf9fee0026541c90a6f557a3e0bc", GitTreeState:"clean"}
```

Output of `kubectl version`:

(N.B.: We are using Kubernetes `v1.7.4`)

```
Client Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.4+0c1a5fc", GitCommit:"0c1a5fc302a1f849e245e64f3ec0d71fb54df8a7", GitTreeState:"not a git tree", BuildDate:"2017-09-06T14:42:45Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.4+0c1a5fc", GitCommit:"0c1a5fc302a1f849e245e64f3ec0d71fb54df8a7", GitTreeState:"not a git tree", BuildDate:"2017-09-06T14:42:45Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by technosophos** [comment]
If --dry-run shows the field modified, then it is likely that Kubernetes dropped the field silently (which is the expected behavior for some immutable fields, such as those on deployments.).

Does using `helm upgrade --force` fix that? This will destroy and recreate the resource, rather than sending a patch operation to Kubernetes.

**Comment by jaytaylor** [comment]
I see, this is helpful information @technosophos - thank you very much!

I'll try with `--force` and return with findings.

**Comment by jaytaylor** [comment]
There was no change in behavior with `--force`, initContainers remain stuck at the previous version.

**Comment by jaytaylor** [comment]
Perusing the code a little, I noticed `https://github.com/helm/helm/blob/master/pkg/tiller/release_update.go#L145`:

```go
// performUpdateForce performs the same action as a `helm delete && helm install --replace`.
```

Thinking through this a little- deletion will be a problem, since resources deletion leads to the associated deployment strategies not being enforced.

Any ideas on how to keep the `deploymnet.initContainer.image` values to stay in sync with `delpoyment.container.image`?  I want them to always match.

**Comment by jaytaylor** [comment]
Perusing the code a little, I noticed [/master/pkg/tiller/release_update.go](https://github.com/helm/helm/blob/master/pkg/tiller/release_update.go#L145):

```go
// performUpdateForce performs the same action as a `helm delete && helm install --replace`.
```

Thinking through this a little- deletion will be a problem, since resources deletion leads to the associated deployment strategies not being enforced.

Any ideas on how to keep the `deploymnet.initContainer.image` values to stay in sync with `delpoyment.container.image`?  I want them to always match.

**Comment by technosophos** [comment]
Can you get the desired behavior to work by doing a `kubectl edit` of the deployment?

**Comment by jaytaylor** [comment]
Yes, I tried that yesterday- to no avail.

It seems strange of kubernetes to disallow any modifications of initContainer values.  The expectation that if a `container.image` is editable, that an `initContainer.image` would also be editable seems like a reasonable one.  Pretty sure I'm barking up the wrong tree to bring it up with this audience in this forum, though.

One odd thing about the helm-end of things is that `helm upgrade --install --force` doesn't trigger the resources to be deleted and re-created in this case.

**Comment by jaytaylor** [comment]
### _Good news!_

I've found a workaround which is easy to implement and does not have any serious drawbacks.

## Background

Starting around the time of Kubernetes v1.7, several issues were filed saying that [initContainer changes were being ignored](https://github.com/kubernetes/kubernetes/issues/47264).  Once [Tom Xing](https://github.com/kubernetes/kubernetes/issues/47264#issuecomment-311584187) dug in, the root cause was identified.  In short, internal object conversions clobber the field updates.

The workaround below is verified compatible with Kubernetes v1.7 and v1.8.

According to the referenced thread, for k8s versions v1.9 or newer, the `pod.beta.kubernetes.io/init-containers` annotation was slated to be deprecated.  I don't presently have a 1.9 cluster available to test these characteristics:

1. If this issue is fixed in 1.9
2. If the beta annotation still works or not in 1.9

_Future readers: If you are able to test these, please reply and add your experience and/or findings!_

## Workaround Details

The technique to sidestep this bug uses the `pod.beta.kubernetes.io/init-containers` annotation + JSON for the initContainer definitions.  Defining the initContainers in this alternative facility causes Kubernetes to process them using a different code path.  Most importantly, a path which doesn't clobber the fields and expose you to this unfortunate bug!

I've verified that initContainers defined this way are able to have the `initContainer.image` field modified by subsequent helm chart upgrades.

### Example

```yaml
spec:
  template:
    metadata:
      annotations:
        pod.beta.kubernetes.io/init-containers: '[{"name": "volume-mount-hack", "image": "my-image:1.2.3", "imagePullPolicy": "Always", "securityContext": {"runAsUser": 0, "command": ["/bin/sh", "-c", "set -e && echo \"[ $(hostname --fqdn) ] Chowning volume mounts to 1000:1000 ..\" && set -x && chown -R 1000:1000 /data /var/log/gigawatt.io"], "volumeMounts": [{"name": "storage", "mountPath": "/data"}, {"name": "log-dir", "mountPath": "/var/log/gigawatt.io"}]}]'
```

Hopefully this is information and trail of breadcrumbs will be useful to some folks in the future.

---

P.S. @technosophos Thank you for creating such a nice and cool way to manage kubernetes workloads!
